chore: add bundle size analysis tools

adriandarian opened the pull request:
## Summary

This PR adds bundle-size monitoring and visualization for OpenSeadragon.

- Adds `npm run bundlesize`, backed by `scripts/check-bundle-size.js`, to check the minified bundle against the limits configured in `package.json`.
- Adds `npm run bundlesize:visualize` using `source-map-explorer` to inspect what contributes to the generated bundle.
- Adds a `Bundle Size Check` GitHub Actions workflow that builds the PR branch, runs the bundle-size check, compares against the base branch, and posts or updates a PR comment with the raw check output plus size deltas.
- Updates `CONTRIBUTING.md` to point contributors at `npm run bundlesize` without duplicating limit values that can drift from `package.json`.

## Usage

```bash
npm run build
npm run bundlesize
npm run bundlesize:visualize
```

## Verification

Local checks:

```bash
npm run build
npm run bundlesize
npx tsc --noEmit -p tsconfig.dts.json
node --check scripts/check-bundle-size.js
node node_modules/source-map-explorer/bin/cli.js --version
```

GitHub Actions:

- `Bundle Size Check / Check Bundle Size` passed.
- `Pull Request Checks / Run test tasks` passed.

iangilman reviewed (commented):
Keeping an eye on our bundle size seems like a good idea; thank you for getting this rolling.

I checked out this branch and did an `npm install` and tried to use the new commands, but none of them work as described. Have you tested this patch?

Having a visualization of what contributes to the bundle size sounds amazing... I'd love to see it! Please let me know how to make that happen.

iangilman reviewed (changes requested):
Okay, this is working great now! Looks good to me. Just a couple of comments, and also please fix the merge conflict.

adriandarian commented:
@iangilman all patched up and CI is passing again

```markdown
Run npm run bundlesize

> openseadragon@6.0.0 bundlesize
> bundlesize

 PASS  build/openseadragon/openseadragon.min.js: 79.78KB < maxSize 85KB (gzip) 

 PASS  build/openseadragon/openseadragon.min.js: 323.32KB < maxSize 330KB (no compression) 
 ```

iangilman reviewed (changes requested):
Thank you for bringing this up to date. Apologies for not getting to it sooner. I'm holding off on your other PRs until we get v6 released, but I realized this one is fine to do now.

Just a few more notes, and yet another merge conflict, but we're basically there on this one... thank you for your patience!

iangilman commented:
@adriandarian Looks like this one is basically done if you want to finish it up!

One additional thought (beyond the notes immediately above and the merge conflicts): I believe the CI action could be updated to print the result of the bundle size check directly into any PR it runs on. If this is easy, maybe add it to this patch? Otherwise no need to drag this out further…

adriandarian committed:
chore: add bundle size analysis tools

adriandarian commented on CONTRIBUTING.md:
```diff
@@ -55,6 +55,23 @@ You can also publish the built version to the site-build repository. This assume
+
+To ensure OpenSeadragon stays lean and fast, we track bundle sizes and enforce limits. To check bundle size:
+
+    npm run analyze
```
Done. `CONTRIBUTING.md` now says `npm run bundlesize`.

adriandarian commented on CONTRIBUTING.md:
```diff
@@ -55,6 +55,23 @@ You can also publish the built version to the site-build repository. This assume
+    npm run analyze
+
+This will show you the current bundle sizes and whether they're within acceptable limits:
+- **Minified (gzipped)**: Must be under 80 KB
```
Done. I removed the hard-coded size numbers from `CONTRIBUTING.md`; it now points readers to the limits configured in `package.json` so the docs do not drift.

adriandarian commented:
@iangilman this is updated and ready for another look.

- Rebasing/merge conflict is resolved cleanly: `adriandarian:feat/bundle-size-analyzer` is now directly on top of current `openseadragon:master`.
- The PR diff is back down to the bundle-size-related files.
- `CONTRIBUTING.md` now uses `npm run bundlesize` and no longer duplicates size-limit values.
- The bundle-size workflow now includes the raw `npm run bundlesize` output in the PR comment.
- Both GitHub checks are passing: `Bundle Size Check / Check Bundle Size` and `Pull Request Checks / Run test tasks`.

iangilman commented:
@adriandarian Thank you for updating this! And thank you for adding the PR commenting. That will be great to have!

Looks like Copilot found some things worth addressing (though probably not all of its comments are important); please take a look.

adriandarian committed:
Merge branch 'master' into feat/bundle-size-analyzer

adriandarian committed:
fix: harden bundle size reporting

adriandarian commented:
Followed up on the new Copilot review in `f5de7bc6`.

Addressed the substantive workflow/checker issues:

- removed the indented multi-line comment template so the PR comment renders as Markdown instead of a code block
- moved `github-script` inputs to `env` and parse them inside the script
- made byte formatting consistent with the local bundle-size checker
- fixed negative and zero-base diff handling
- switched the workflow timestamp to ISO format
- added a stable HTML marker for bot comment updates
- moved the saved `bundlesize` output into `RUNNER_TEMP`
- added `package-lock.json` to the workflow path filter
- removed the unnecessary global `grunt-cli` install
- tightened `parseMaxSize` in `scripts/check-bundle-size.js`

Local verification:

```bash
npm run build
npx tsc --noEmit -p tsconfig.dts.json
npm run bundlesize
```

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+              const k = 1024;
+              const sizes = ['B', 'KB', 'MB'];
+              const i = Math.floor(Math.log(bytes) / Math.log(k));
+              return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
```
Addressed in `f5de7bc6`. The workflow formatter now uses decimal units (`1000`, `kB`/`MB`) to match `scripts/check-bundle-size.js`, so the table output and embedded `npm run bundlesize` output use the same convention.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+            const baseMinSize = parseInt('${{ steps.base_size.outputs.base_min_size }}');
+            const baseFullSize = parseInt('${{ steps.base_size.outputs.base_full_size }}');
+            const baseMinGzip = parseInt('${{ steps.base_size.outputs.base_min_gzip }}');
+            const baseFullGzip = parseInt('${{ steps.base_size.outputs.base_full_gzip }}');
```
Addressed in `f5de7bc6`. The `github-script` step now receives the size values through `env` and parses them from `process.env` inside the script instead of interpolating `${{ ... }}` directly into the JavaScript body.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+              const diffPercent = ((diff / base) * 100).toFixed(2);
+              const sign = diff > 0 ? '+' : '';
+              const direction = diff > 0 ? 'increased' : diff < 0 ? 'decreased' : 'unchanged';
+              return `${direction}: ${sign}${formatBytes(diff)} (${sign}${diffPercent}%)`;
```
Addressed in `f5de7bc6`. `formatDiff` now guards `base === 0` and reports `n/a` for the percentage in that edge case instead of producing `NaN%` or `Infinity%`.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+name: 'Bundle Size Check'
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+    paths:
+      - 'src/**'
+      - 'Gruntfile.js'
+      - 'package.json'
+      - '.github/workflows/bundle-size.yml'
+
+jobs:
+  check_bundle_size:
+    name: Check Bundle Size
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      pull-requests: write
+      issues: write
```
Addressed in `f5de7bc6`. I left the workflow on `pull_request` for the safer token model, and I updated the catch-path logging to explicitly say comment writes can fail for fork PRs because of read-only permissions.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+          set -o pipefail
+          npm run bundlesize 2>&1 | tee bundlesize-output.txt
+
+      - name: Checkout base branch
+        run: |
+          git checkout ${{ github.base_ref }}
+
+      - name: Clean and rebuild base branch
+        run: |
+          npm ci
+          npm run build
```
Addressed in `f5de7bc6`. The PR branch bundlesize output is now written to `$RUNNER_TEMP/bundlesize-output.txt` before switching branches, so it no longer relies on an untracked repo file surviving the checkout.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+            ---
+            <sub>Measured on ${new Date().toLocaleString()}</sub>
+            `;
+
```
Addressed in `f5de7bc6`. The PR comment body is now built from a newline-joined array with no leading indentation, so the tables/headings/fenced block render as normal Markdown.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+
+      - name: Install dependencies (PR branch)
+        run: |
+          npm install -g grunt-cli
```
Addressed in `f5de7bc6`. I removed the global `npm install -g grunt-cli`; the workflow just runs `npm ci` and then `npm run build`.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+            ${minGzipDiff > 0 ? '**Warning**: Bundle size increased.' : minGzipDiff < 0 ? '**Good**: Bundle size decreased.' : '**Good**: No change in bundle size.'}
+
+            ---
+            <sub>Measured on ${new Date().toLocaleString()}</sub>
```
Addressed in `f5de7bc6`. The timestamp now uses `toISOString()` so the PR comment gets a stable, unambiguous UTC timestamp.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+    paths:
+      - 'src/**'
+      - 'Gruntfile.js'
+      - 'package.json'
```
Addressed in `f5de7bc6`. `package-lock.json` is now included in the workflow `paths:` filter so dependency-only lockfile changes still trigger the bundle-size check.

adriandarian commented on scripts/check-bundle-size.js:
```diff
@@ -0,0 +1,76 @@
+const path = require('path');
+const zlib = require('zlib');
+
+const root = path.resolve(__dirname, '..');
+const packageJson = require(path.join(root, 'package.json'));
+const checks = packageJson.bundlesize || [];
+
+function parseMaxSize(size) {
+    const match = /^([\d.]+)\s*(b|kb|mb)$/i.exec(size);
+
+    if (!match) {
+        throw new Error(`Unsupported maxSize value: ${size}`);
+    }
+
+    const amount = Number(match[1]);
+    const unit = match[2].toLowerCase();
+    const multipliers = {
+        b: 1,
+        kb: 1000,
+        mb: 1000 * 1000
+    };
+
+    return amount * multipliers[unit];
+}
```
Addressed in `f5de7bc6`. `parseMaxSize` now uses a stricter decimal regex and also validates the parsed number with `Number.isFinite` before applying the unit multiplier.

adriandarian commented on scripts/check-bundle-size.js:
```diff
@@ -0,0 +1,76 @@
+    return amount * multipliers[unit];
+}
+
+function formatSize(bytes) {
+    if (bytes < 1000) {
+        return `${bytes}B`;
+    }
+
+    return `${(bytes / 1000).toFixed(2)}KB`;
+}
```
Addressed in `f5de7bc6`. `formatSize` now reports decimal units as `kB`, matching the workflow formatter so the local checker output and the PR comment table stay aligned.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,198 @@
+                issue_number: context.issue.number,
+              });
+
+              const botComment = comments.find(comment =>
+                comment.user.type === 'Bot' && comment.body.includes('Bundle Size Report')
+              );
```
Addressed in `f5de7bc6`. Bot comment detection now uses a dedicated HTML marker (`<!-- bundle-size-report -->`) instead of matching any bot comment that happens to contain `Bundle Size Report`.

iangilman commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,221 @@
+
+            try {
+              // Find existing comment
+              const { data: comments } = await github.rest.issues.listComments({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                issue_number: context.issue.number,
+              });
+
+              const botComment = comments.find(comment =>
+                comment.user.type === 'Bot' && comment.body.includes(marker)
+              );
```
Actually, now that I think about it, I wonder if it's better to have it post a new comment each time? I guess that would be 1 per push? That way it is easy to see the change even as the PR evolves.

Anyway, maybe we just leave this as it is and see how it feels.

adriandarian replied:
Leaving this one as-is for now. The current update-in-place behavior keeps a single bundle-size report on the PR, and given your note about possibly preferring a fresh comment per push, I did not add pagination logic in this pass.

adriandarian replied:
can raise as new pr after this merge if it turns out you prefer multi comment

iangilman reviewed (approved):
Looks great to me! Thank you for addressing all of Copilot's comments.

This next set all seems pretty minor. I'm happy to merge without addressing them, but I'll give you a chance to make any changes if you want to.

adriandarian committed:
fix: harden bundle size workflow pipelines

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,221 @@
+
+              const k = 1000;
+              const sizes = ['B', 'kB', 'MB'];
+              const i = Math.floor(Math.log(absoluteBytes) / Math.log(k));
```
Addressed in `e88c794d`. `formatBytes` now clamps the unit index to the end of the unit list, so large values still render predictably instead of indexing past the available suffixes.

adriandarian commented on .github/workflows/bundle-size.yml:
```diff
@@ -0,0 +1,221 @@
+        run: npm run build
+
+      - name: Get PR branch bundle sizes
+        id: pr_size
+        run: |
+          # Get file sizes
+          PR_MIN_SIZE=$(stat -f%z "build/openseadragon/openseadragon.min.js" 2>/dev/null || stat -c%s "build/openseadragon/openseadragon.min.js")
+          PR_FULL_SIZE=$(stat -f%z "build/openseadragon/openseadragon.js" 2>/dev/null || stat -c%s "build/openseadragon/openseadragon.js")
+
+          # Get gzipped sizes
+          PR_MIN_GZIP=$(gzip -c "build/openseadragon/openseadragon.min.js" | wc -c | tr -d ' ')
+          PR_FULL_GZIP=$(gzip -c "build/openseadragon/openseadragon.js" | wc -c | tr -d ' ')
+
```
Addressed in `e88c794d`. I added `set -o pipefail` to both bundle-size collection steps so a gzip failure cannot be masked by the downstream `wc -c | tr -d ' '` pipeline.

iangilman commented:
Great, thank you for those last fixes!
